src/main/java/pvs/app/service/GitlabApiService.java

```diff
@@ -0,0 +1,133 @@
+        String token = System.getenv("PVS_GITLAB_TOKEN");
+        if (useToken){
+            this.gitLabApi = new GitLabApi("https://gitlab.com", Constants.TokenType.ACCESS, token);
+        }
+
+        this.gitlabCommitService = gitlabCommitService;
+        this.webClient = webClientBuilder.baseUrl("https://api.gitlab.com/")
+                .defaultHeader("Authorization", "Bearer " + token)
+                .build();
+    }
+
+    private String dateToISO8601(Date date) {
+        SimpleDateFormat sdf;
+        sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
+        sdf.setTimeZone(TimeZone.getTimeZone("GMT+8"));
+        return sdf.format(date);
+    }
+
+    public void oauth2login(String username, String password) throws GitLabApiException {
+        if (this.gitLabApi == null) {
+            this.gitLabApi = GitLabApi.oauth2Login("https://gitlab.com/", username, password);
+        }
+        System.out.println("oauth2login has been executed");
+        // two problems need to be dealt with:
+        // 1. password security
+        // 2. whether we need to log in whenever the service start
+    }
+
+    public Boolean getCommitsFromGitlab(String owner, String name) throws GitLabApiException, InterruptedException {
+        CommitsApi commitsApi = this.gitLabApi.getCommitsApi();
+        ProjectApi projectApi = this.gitLabApi.getProjectApi();
+        System.out.println(owner + " " + name);
+        Project project = projectApi.getProject(owner, name);
+        System.out.println("test1");
+        commits = commitsApi.getCommits(project.getId());
+        if (commits.size() != 0){
+            List<CommitStats> commitStats = new ArrayList<>();
+            for(Commit commit: commits){
+                commitStats.add(commitsApi.getCommit(project.getId(),commit.getId()).getStats());
+            }
+
+            List<GitlabCommitLoaderThread> gitlabCommitLoaderThreadList = new ArrayList<>();
+
+            for (int i = 0; i < commits.size(); i++) {
+                GitlabCommitLoaderThread gitlabCommitLoaderThread =
+                        new GitlabCommitLoaderThread(
+                                this.gitlabCommitService,
+                                this,
+                                owner, // repoOwner
+                                name,  // repoName
+                                String.valueOf(commits.get(i)), // commit info
+                                commitStats.get(i),  // commit stats
+                                commitsApi.getDiff(project.getId(), commits.get(i).getId()).size() // changeFileCount
+                        );
+                gitlabCommitLoaderThreadList.add(gitlabCommitLoaderThread);
+                gitlabCommitLoaderThread.start();
+            }
+
+            for (GitlabCommitLoaderThread thread : gitlabCommitLoaderThreadList) {
+                thread.join();
+            }
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
Suggest use early-return.
And also suggest unwrap `else { ... }` in other places if there's already have a `return ...` above.

```java
        if (commits.size() == 0) return false;

        final List<CommitStats> commitStats = new ArrayList<>();

        for (final Commit commit : commits) {
            commitStats.add(commitsApi.getCommit(project.getId(), commit.getId()).getStats());
        }

        final List<GitlabCommitLoaderThread> gitlabCommitLoaderThreadList = new ArrayList<>();

        for (int i = 0; i < commits.size(); i++) {
            final GitlabCommitLoaderThread gitlabCommitLoaderThread =
                    new GitlabCommitLoaderThread(
                            this.gitlabCommitService,
                            owner, // repoOwner
                            name,  // repoName
                            String.valueOf(commits.get(i)), // commit info
                            commitStats.get(i),  // commit stats
                            commitsApi.getDiff(project.getId(), commits.get(i).getId()).size() // changeFileCount
                    );
            gitlabCommitLoaderThreadList.add(gitlabCommitLoaderThread);
            gitlabCommitLoaderThread.start();
        }

        for (final GitlabCommitLoaderThread thread : gitlabCommitLoaderThreadList) {
            thread.join();
        }

        return true;
```